Catalogue Index Optimization

danieldresser-ie opened the pull request:
This is a simple initial implementation of the catalogue optimization we talked about.  Definitely needs a bit more interactive testing, and validation that we're actually seeing the performance gains we want, but it feels about right, and the tests are passing.

One thing to think about:  it would have been easy to stick "output:*" in the same map to accelerate those as well ( perhaps even easier than having a separate branch ), but then the map would be invalidated whenever an output index changes - assuming that happens more often than a rename, that would be undesirable.

johnhaddon commented:
> Definitely needs a bit more interactive testing, and validation that we're actually seeing the performance gains we want, but it feels about right

I'm not seeing any appreciable performance improvement here. Here's my test scenario :

1. Make a Catalogue with about 1000 images.
2. Scroll to the bottom.
3. Click around in the output column to assign index 1 to different images.
4. Notice the lag between the click and the status update.
5. Run VTune. Notice that the majority of time is spent in hashing.

I think we still need to remove `g_imageNameContextName` from the context before accessing `imageIndexMapPlug()`, otherwise we still have quadratic hashing performance. I tried that here and it seemed to do the trick. There's still a bit of lag, but it's appreciably less, and VTune indicates all the time is in Python stuff now, which is the catalogue columns themselves (I have a plan for improving this part).

> One thing to think about: it would have been easy to stick "output:*" in the same map to accelerate those as well ( perhaps even easier than having a separate branch )

That does seem appealing.

> but then the map would be invalidated whenever an output index changes - assuming that happens more often than a rename, that would be undesirable

The case I'm most interested in is the output index changing, because any lag between clicking and seeing the new status feels very off. But if we're going from `N*N` to just `N`, it seems like it'll probably be fine anyway? Worth trying I think...

I noticed a separate small regression when testing this : an empty Catalogue no longer renders as an empty image. Instead it shows a message saying `Unknown Image ""`. Could you fix that too please?

danieldresser-ie commented:
I've removed the image name from the context, and validated that this does improve performance.  I've also put the output indices into the map, and that seems to have no perceptible impact on performance when switching outputs, so it's probably preferable ( it should definitely have better performance when outputs aren't being switched ) ... I've left this in a separate commit to review before squashing, so this isn't ready to merge quite yet ( I'll add Changelog entries at the same time ).

I'm testing on a 2048 image catalog, and there is still some noticeable slowdown when switching outputs even after this change, but it seems to all be coming from PathListingWidget.updateColumnWidths, so seems unrelated to this code.

For the regression for empty catalogs, I've just added a case when computing invalidImageTextPlug() - not sure if it feels messy to now be using a Text node in this case, but the result is the same, and it's much simpler than having to disable things or add a special switch input in this one case.

danieldresser-ie commented:
Squashed and added Changelog entries.
